controllers/topology/internal/scope/upgradetracker.go

```diff
@@ -22,31 +22,68 @@ const maxMachineDeploymentUpgradeConcurrency = 1
 
 // UpgradeTracker is a helper to capture the upgrade status and make upgrade decisions.
 type UpgradeTracker struct {
+	ControlPlane       ControlPlaneUpgradeTracker
 	MachineDeployments MachineDeploymentUpgradeTracker
 }
 
+// ControlPlaneUpgradeTracker holds the current upgrade status of the Control Plane.
+type ControlPlaneUpgradeTracker struct {
+	IsAtTopologyVersion bool
+	IsProvisioning      bool
+	IsUpgrading         bool
+	IsScaling           bool
+}
+
 // MachineDeploymentUpgradeTracker holds the current upgrade status and makes upgrade
 // decisions for MachineDeployments.
 type MachineDeploymentUpgradeTracker struct {
-	names sets.String
+	pendingUpdate     sets.String
+	startRollOutNames sets.String
+	RollingOutNames   []string
 }
 
 // NewUpgradeTracker returns an upgrade tracker with empty tracking information.
 func NewUpgradeTracker() *UpgradeTracker {
 	return &UpgradeTracker{
 		MachineDeployments: MachineDeploymentUpgradeTracker{
-			names: sets.NewString(),
+			startRollOutNames: sets.NewString(),
+			pendingUpdate:     sets.NewString(),
 		},
 	}
 }
 
+// AllRollingOutNamesList returns the list of machine deployments that are rolling out and
+// are about to roll out.
+func (m *MachineDeploymentUpgradeTracker) AllRollingOutNamesList() []string {
+	return append(m.startRollOutNames.List(), m.RollingOutNames...)
+}
+
 // Insert adds name to the set of MachineDeployments that will be upgraded.
 func (m *MachineDeploymentUpgradeTracker) Insert(name string) {
-	m.names.Insert(name)
+	m.startRollOutNames.Insert(name)
 }
 
 // AllowUpgrade returns true if a MachineDeployment is allowed to upgrade,
 // returns false otherwise.
 func (m *MachineDeploymentUpgradeTracker) AllowUpgrade() bool {
-	return m.names.Len() < maxMachineDeploymentUpgradeConcurrency
+	return m.startRollOutNames.Len() < maxMachineDeploymentUpgradeConcurrency
+}
+
+// MarkPendingUpdate marks a machine deployment as in need of an update.
+// This is generally used to capture machine deployments that have not been
+// updated to the topology version.
+func (m *MachineDeploymentUpgradeTracker) MarkPendingUpdate(name string) {
+	m.pendingUpdate.Insert(name)
+}
+
+// HasPendingUpdates returns true if any of the machine deployments are still
+// waiting to pick up the topology version. Returns false, otherwise.
+func (m *MachineDeploymentUpgradeTracker) HasPendingUpdates() bool {
```

Comment:
Might be I'm missing some detail but it seems to me we are using rollout, upgrade, and updates in a mixed way. Can we converge on one/simplify naming? Otherwise let's provide comments explaining the different meanings

Reply:
I renames the variable and function names to make it more consistent.